odo/backends/sql.py

```diff
@@ -199,8 +199,10 @@ def discover_sqlalchemy_column(c):
 
 @discover.register(sa.sql.FromClause)
 def discover_sqlalchemy_selectable(t):
-    ordering = dict((c, i) for i, c in enumerate(c for c in t.columns.keys()))
+    ordering = dict((str(c), i)
+                    for i, c in enumerate(c for c in t.columns.keys()))
     records = list(sum([discover(c).parameters[0] for c in t.columns], ()))
+    records = [(str(r[0]), r[1]) for r in records]
```

Comment:
The original expression is a little confusing, and iterating over the sequence again is a bit unclear. It took me a while to walk through the intent of this code, mostly because of what already existed though. Do you think we could pull this into another helper function? something like:

```python
def _process_columns(columns):
    """Process the dshapes of the columns of a table.

    Parameters
    ----------
    columns : iterable[column]
        The columns to process.

    Yields
    ------
    record_entry : tuple[str, dshape]
        A record entry containing the name and type of each column.
    """
    for col in columns:
        (name, dtype), = discover(col).fields
        yield str(name), dtype
```
here fields replaces parameters[0] because we know this should be a record type.

We can then replace this call site with something like:

```python
record = list(_process_columns(t.columns))
```

Also, we should probably call this `record` instead of `records` because it is the types for a single record and thus not plural.

I think this expression would be more clear but I am happy to hear what you think.